File: W3D4-merge_sort/merge_sort.rb
class Array

  def merge_sort(&prc)
    return self if self.length == 1

    prc ||= Proc.new {|ele1, ele2| ele1 <=> ele2}

    mid_idx = self.length / 2
    left_arr = self[0...mid_idx]
        # when self.length == 2 -> 2 / 2 = 1 -> 0..1
    right_arr = self[mid_idx..-1]

    left_call = left_arr.merge_sort(&prc)
    right_call = right_arr.merge_sort(&prc)

    Array.merge(left_call, right_call, &prc)
  end

  private
  def self.merge(left_call, right_call, &prc)
    merged = []
    while !left_call.empty? && !right_call.empty?
        if prc.call(left_call[0], right_call[0]) != 1
            merged << left_call.shift
        else
            merged << right_call.shift
        end
    end
    merged + left_call + right_call
  end
end


p [4, 3, 8, 0, 5, 6].merge_sort() # => [0, 3, 4, 5, 6, 8]
p [4, 3, 8, 0, 5, 6].merge_sort() {|a, b| b <=> a} # => [8, 6, 5, 4, 3, 0]

File: W8D3-mergeSort/mergeSort.js
Array.prototype.mergeSort = function (callback) {
    if (this.length < 2) return this;
    let midIdx = this.length / 2;

    // leftCall = this.slice(0, midIdx).mergeSort(callback);
    let leftArray = this.slice(0, midIdx);
    let rightArray = this.slice(midIdx);
    let leftCall = leftArray.mergeSort(callback);
    let rightCall = rightArray.mergeSort(callback);
    
    return merge(leftCall, rightCall, callback)
};


function merge(left, right, callback) {
    let merged = []
    while ((left.length !== 0) && (right.length !== 0)) {
        if (callback(left[0], right[0]) === -1) {
            merged.push(left.shift())
        } else {
            merged.push(right.shift())
        }
    }
    return merged.concat(left, right)
};

function revCb (x, y) {
    if (x === y) {
        return 0;
    } else if (x < y) {
        return 1;
    } else {
        return -1;
    }
}

function normCb(x, y) {
    if (x === y) {
        return 0;
    } else if (x > y) {
        return 1;
    } else {
        return -1;
    }
}

console.log([5, 4, 7, 9, 3, 6, 0].mergeSort(revCb)) //[ 9, 7, 6, 5, 4, 3, 0 ]
console.log([5, 4, 7, 9, 3, 6, 0].mergeSort(normCb)) // [ 0, 3, 4, 5, 6, 7, 9 ]

// console.log(['a', 'b', 'c'].concat(['a'], ['d', 'e', 'f'])) ==> ["a", "b", "c", "a", "d", "e", "f"]